File: python/app/logship.py
"""
Логи python-analytics → stdout + Logstash/ELK.

Тот же контракт, что у Go-API (go/internal/logship), Nitro-роутов Nuxt
(nuxt/server/utils/logstash.ts) и раздела «Себестоимость»
(python/cost/app/logship.py):

  * одна JSON-запись на строку (кодек json_lines), поля
    time/level/msg/service/host + свои — чтобы в общем индексе ELK все рантаймы
    кабинета лежали единообразно. service=finance-analytics;
  * базовый канал — stdout (его собирает docker) — остаётся ВСЕГДА;
  * отправка асинхронная и НИКОГДА не блокирует обработку запроса: строка
    кладётся в очередь, фоновый поток-демон держит TCP-соединение и
    переподключается с backoff. Нет связи или очередь переполнена — строка
    отбрасывается, растёт счётчик dropped (его раз в минуту пишет сам логгер).

Копия, а не импорт из python/cost/: контуры изолированы и собираются в разные
образы (swarm/python-analytics/Dockerfile.dev видит только python/, образ
python-cost — только python/cost/). Зависимостей у модуля нет — чистый stdlib,
как и весь этот sandbox.

ENV:
  LOGSTASH_HOST — пусто → трансляция выключена, остаётся только stdout;
  LOGSTASH_PORT — пусто при заданном HOST → 5044.
"""
from __future__ import annotations

import json
import logging
import os
import queue
import socket
import threading
import time
from datetime import datetime, timezone
from typing import Any

SERVICE = "finance-analytics"
HOSTNAME = socket.gethostname()

_HOST = os.environ.get("LOGSTASH_HOST", "").strip()
_PORT = int(os.environ.get("LOGSTASH_PORT") or "5044")
_MAX_QUEUE = 4096
_DIAL_TIMEOUT = 5.0
_BACKOFF_START = 1.0
_BACKOFF_MAX = 30.0

# Python зовёт уровни иначе, чем slog/Nitro — приводим к общему словарю.
_LEVELS = {"WARNING": "WARN", "CRITICAL": "ERROR"}


class _Shipper:
    """Фоновый поток-демон: очередь строк → TCP-соединение с Logstash."""

    def __init__(self, host: str, port: int) -> None:
        self._host = host
        self._port = port
        self._queue: queue.Queue[str] = queue.Queue(maxsize=_MAX_QUEUE)
        self.dropped = 0
        threading.Thread(target=self._run, name="logship", daemon=True).start()

    def ship(self, line: str) -> None:
        """Кладёт строку (с '\\n' на конце) в очередь. Не блокирует, не бросает."""
        try:
            self._queue.put_nowait(line)
        except queue.Full:
            self.dropped += 1  # очередь полна — роняем, не тормозим запрос

    def _run(self) -> None:
        sock: socket.socket | None = None
        backoff = _BACKOFF_START
        while True:
            line = self._queue.get()
            if sock is None:
                try:
                    sock = socket.create_connection((self._host, self._port), _DIAL_TIMEOUT)
                except OSError:
                    # Logstash недоступен — роняем строку и ждём backoff, чтобы
                    # не долбить мёртвый адрес на каждой записи лога.
                    self.dropped += 1
                    time.sleep(backoff)
                    backoff = min(backoff * 2, _BACKOFF_MAX)
                    continue
                backoff = _BACKOFF_START
            try:
                sock.sendall(line.encode("utf-8"))
            except OSError:
                try:
                    sock.close()
                finally:
                    sock = None
                self.dropped += 1  # строку потеряли, переподключимся на следующей


_shipper: _Shipper | None = None


def ship_log(line: str) -> None:
    if _shipper is not None:
        _shipper.ship(line)


def dropped_count() -> int:
    return _shipper.dropped if _shipper is not None else 0


class JsonFormatter(logging.Formatter):
    """Формат записи — общий для всех рантаймов кабинета (см. шапку модуля)."""

    def format(self, record: logging.LogRecord) -> str:  # noqa: A003
        payload: dict[str, Any] = {
            "time": datetime.fromtimestamp(record.created, timezone.utc)
            .isoformat()
            .replace("+00:00", "Z"),
            "level": _LEVELS.get(record.levelname, record.levelname),
            "msg": record.getMessage(),
            "service": SERVICE,
            "host": HOSTNAME,
            "logger": record.name,
        }
        fields = getattr(record, "fields", None)
        if isinstance(fields, dict):
            payload.update(fields)
        if record.exc_info:
            payload["error"] = self.formatException(record.exc_info)
        return json.dumps(payload, ensure_ascii=False, default=str)


class _StdoutAndLogstashHandler(logging.StreamHandler):
    """stdout — базовый канал; та же строка уходит в Logstash (если включён)."""

    def emit(self, record: logging.LogRecord) -> None:
        super().emit(record)
        try:
            ship_log(self.format(record) + "\n")
        except Exception:  # noqa: BLE001 — лог не имеет права ронять запрос
            pass


def log(level: int, msg: str, **fields: Any) -> None:
    """Разовая структурная запись: log(logging.INFO, "…", rows=123)."""
    logging.getLogger("analytics").log(level, msg, extra={"fields": fields})


def _watch_dropped() -> None:
    """Раз в минуту сообщает, сколько строк не доехало в Logstash.

    Рост счётчика = канал в ELK ослеп (нет связи либо переполняется очередь);
    это повод для алерта. Пишем только при изменении, чтобы не сорить.
    """
    last = 0
    while True:
        time.sleep(60)
        now = dropped_count()
        if now != last:
            log(logging.WARNING, "logship: строки лога не доставлены в Logstash",
                dropped_total=now, since_last=now - last)
            last = now


def setup_logging() -> None:
    """Переводит логирование процесса на JSON и поднимает канал в Logstash."""
    global _shipper
    if _HOST and _shipper is None:
        _shipper = _Shipper(_HOST, _PORT)

    handler = _StdoutAndLogstashHandler()
    handler.setFormatter(JsonFormatter())

    root = logging.getLogger()
    root.handlers = [handler]
    root.setLevel(logging.INFO)

    if _shipper is not None:
        log(logging.INFO, "logship: трансляция логов в Logstash включена",
            addr=f"{_HOST}:{_PORT}")
        threading.Thread(target=_watch_dropped, name="logship-dropped", daemon=True).start()

File: go/internal/logship/logship.go
// Package logship — асинхронная трансляция структурных логов (slog JSON) в
// Logstash по TCP, кодек json_lines. Writer встаёт ВТОРЫМ в io.MultiWriter
// рядом с os.Stdout: stdout остаётся базовым каналом (его собирает docker),
// ELK — дополнительный. Порт того же решения, что в кабинете NCI.
//
// Инвариант, ради которого всё и написано: Write НИКОГДА не блокирует и НИКОГДА
// не возвращает ошибку. slog зовётся из каждого HTTP-хендлера (access-лог) —
// заблокируйся отправка в недоступный Logstash, и встал бы горячий путь, а через
// MultiWriter умерла бы и запись в stdout. Поэтому Write лишь кладёт копию строки
// в буфер и мгновенно возвращается; фоновая горутина держит TCP-соединение,
// переподключается с backoff и сливает буфер. Переполнен буфер или нет связи —
// строка отбрасывается, а счётчик Dropped растёт (main периодически пишет его
// в stdout: рост означает, что канал в ELK ослеп).
package logship

import (
	"net"
	"sync/atomic"
	"time"
)

const (
	dialTimeout  = 5 * time.Second
	writeTimeout = 5 * time.Second
	backoffStart = 1 * time.Second
	backoffMax   = 30 * time.Second
)

// Writer — асинхронный шиппер строк лога в Logstash. Реализует io.Writer.
type Writer struct {
	addr    string
	queue   chan []byte
	dropped atomic.Int64
	done    chan struct{}
}

// New поднимает шиппер на addr вида "host:port" и запускает фоновую горутину.
// Пустой addr → nil: трансляция выключена, остаётся только stdout (та же
// конвенция «пусто = интеграция отключена», что у SITE_API_NOTIFY_* и
// B24_USERGET_WEBHOOK). Методы на *Writer nil-safe, но io.MultiWriter принимать
// nil нельзя — проверку делает вызывающий код (cmd/api/logging.go).
func New(addr string, queueSize int) *Writer {
	if addr == "" {
		return nil
	}
	w := &Writer{
		addr:  addr,
		queue: make(chan []byte, queueSize),
		done:  make(chan struct{}),
	}
	go w.loop()
	return w
}

// Write кладёт КОПИЮ строки в буфер и возвращается немедленно. Копия
// обязательна: slog переиспользует буфер записи после возврата Write.
// Всегда возвращает (len(p), nil) — см. инвариант в шапке пакета.
func (w *Writer) Write(p []byte) (int, error) {
	if w == nil {
		return len(p), nil
	}
	b := make([]byte, len(p))
	copy(b, p)
	select {
	case w.queue <- b:
	default:
		w.dropped.Add(1) // буфер полон — роняем, не блокируем горячий путь
	}
	return len(p), nil
}

// Dropped — сколько строк потеряно (буфер полон, dial/write не удались).
func (w *Writer) Dropped() int64 {
	if w == nil {
		return 0
	}
	return w.dropped.Load()
}

// Close останавливает фоновую горутину. Строки, оставшиеся в буфере, не
// досылаются — на завершении процесса это приемлемо (базовый stdout не теряем).
func (w *Writer) Close() {
	if w == nil {
		return
	}
	close(w.done)
}

func (w *Writer) loop() {
	var conn net.Conn
	backoff := backoffStart
	defer func() {
		if conn != nil {
			_ = conn.Close()
		}
	}()

	for {
		select {
		case <-w.done:
			return
		case line := <-w.queue:
			if conn == nil {
				c, err := net.DialTimeout("tcp", w.addr, dialTimeout)
				if err != nil {
					// Logstash недоступен — роняем строку и ждём backoff,
					// чтобы не долбить мёртвый адрес на каждой строке.
					w.dropped.Add(1)
					if !w.sleep(backoff) {
						return
					}
					if backoff < backoffMax {
						backoff *= 2
					}
					continue
				}
				conn = c
				backoff = backoffStart
			}
			// Строка slog уже завершается '\n' — ровно то, что ждёт кодек
			// json_lines. Отдельный разделитель добавлять не нужно.
			_ = conn.SetWriteDeadline(time.Now().Add(writeTimeout))
			if _, err := conn.Write(line); err != nil {
				_ = conn.Close()
				conn = nil
				w.dropped.Add(1) // строку потеряли, переподключимся на следующей
			}
		}
	}
}

// sleep ждёт d, но прерывается на Close. Возвращает false, если пора выходить.
func (w *Writer) sleep(d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-w.done:
		return false
	case <-t.C:
		return true
	}
}
